### ISSUE TITLE
Deprecate `twisted.python.util.unsignedID` and replace its uses with `id`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5544]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)| @washort reported|
|-|-|
|Trac ID|trac#5544|
|Type|task|
|Created|2012-03-13 20:47:26Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-unsignedid-5544|

`twisted.python.util.unsignedID` was created to deal with `id` returning negative numbers. Python 2.5 and later don't have this problem and `id` now returns a positive number always.

Replace all uses of this function in Twisted with `id`, then deprecate it.

Attachments:

* [5544.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/640/64016eb15b3176f698926944b44142f973f3f07b/263f6257fd89e51d47c166bf9c2a3192516ffcc9.patch) (13517 bytes) - added by zomux on 2012-04-03 13:35:50Z - 
* [5544.fixed.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/640/64016eb15b3176f698926944b44142f973f3f07b/c07bdeebd455748379dedec3833759980f57754b.patch) (13831 bytes) - added by zomux on 2012-04-13 02:23:15Z - 
* [5544.fixed2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/640/64016eb15b3176f698926944b44142f973f3f07b/a24b9b330fdd2c4e25e000f27e02e14a5fc355ce.patch) (13022 bytes) - added by zomux on 2012-04-17 13:34:39Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5544 5544
type__task task
reporter__washort washort
priority__normal normal
milestone__ 
branch__branches_deprecate_unsignedid_5544 branches/deprecate-unsignedid-5544
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1331671646000000 1331671646000000
changetime__1358561292000000 1358561292000000
version__None None
owner__tom_prince tom.prince
cc__zomux
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="zomux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|zomux set owner to zomux<br>zomux set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="zomux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|zomux removed owner<br>zomux set status to `new`|
|-|-|

Removed twisted.python.util.unsignedID , for deprecated after Python 2.5 
Removed twisted.python.util.setIDFunction , the function used by unsignedID
Removed unit test test_unsignedID
Removed class UnsignedIDTests in twisted.python.test.test_util

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra set owner to zomux|
|-|-|

Thanks for your patch.

 1. This change removes `twisted.python.util.unsignedID` but the [CompatibilityPolicy#HowtodeprecateAPIs compatibility policy] states that it should be deprecated first
 1. `twisted/topfiles/5544.removal` should be one sentence.

**Comment by twisted-trac** [comment]
|<img alt="zomux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|zomux removed owner|
|-|-|

Fixed problems above by thijs :)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to zomux|
|-|-|

1.This patch changes the behaviour of the public `setIDFunction` in `twisted/python/util.py` but should continue to function as before but emit a deprecation warning instead. This also needs to be tested.
 1. The deprecation warning(s) should reference version 12.1.0 instead of 12.0.0
 1. the removal file doesn't flow correctly, what about:
```
twisted.python.util.unsignedID is deprecated, use the built-in function "id" instead
```

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Maybe rather than pointing out all the things wrong with each version of the patch, we should have some documentation about the _right_ way to deprecate something.  Getting to the right solution by random chance seems like it could take a while (and seems like a waste of time).

I created [#5645](https://github.com/twisted/twisted/issues/5645) for documenting this and included at least a partial list of what to do when deprecating a function there.

**Comment by twisted-trac** [comment]
|<img alt="zomux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|zomux removed owner|
|-|-|

fixed problems described by thijs.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to zomux|
|-|-|

Thanks.

  1. `unsignedID` is still imported in `twisted/internet/tcp.py` (as reported by pyflakes)
  1. in `twisted/internet/test/test_base.py`, the old assertion about the repr of a `DelayedCall` had one shortcoming and one advantage relative to the new assertion:
    1. It verified the address in the repr started with "0x".  The new assertion relies on `hex` to return a string starting with this value, which I'd hesitate to rely on (new versions of Python may choose to change this behavior).  Constructing the string using something like "0x%x" is probably better.
    1. The shortened hex string is inconsistent with builtin object reprs.  "0x000000c8" would be a better output (given the old hard-coded id of 200).  A subtlety with fixing this is selecting the right pointer size (32 bit vs 64 bit).  It would be great if you could file a new ticket for fixing this.
  1. Same comment about the change to `twisted/test/test_amp.py`
  1. The unit test for `unsignedID` must not be deleted yet.  The procedure for deprecating a function is to *preserve its behavior* _apart_ from causing it to also emit a warning.  This ensures that applications using the API _continue to work_.  This is the purpose of the deprecation process: to avoid breaking applications based on affected APIs.
  1. Likewise, the implementation of `unsignedID` must therefore not be modified as this patch modifies it.
  1. Nor may `unsignedID` be removed from `__all__` of its defining module.
  1. The change needs a "removal" news fragment, summarizing the deprecation and suggesting the replacement api (the builtin `id` method).

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

> 7. The change needs a "removal" news fragment, summarizing the deprecation and suggesting the replacement api (the builtin id method).

My mistake.  You did add this news file, I wasn't paying close enough attention.  It would be slightly improved by a recommendation to use `id` instead, though.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

(In [36833]) Branching to 'deprecate-unsignedid-5544'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra removed owner|
|-|-|

Please review.

 1. added test for `internet.tcp._BaseTCPClient.__repr__` that was using `util.unsignedID`
 1. removed usage of `util.setIDFunction` (was only used in tests)
 1. deprecated `setIDFunction` and `unsignedID`
 1. removed unused import from `twisted.python.reflect`

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince set owner to @tomprince|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince commented|
|-|-|

  * The tests for the deprecated functions should suppress the warnings from those functions, using something like
```
class Tests(unittest.TestCase):
    suppress  = [((), {'action': ignore', 'message':'%s was deprecated' % (reflect.fullyQualifiedName(unsignedId),), 'category':DeprecationWarning})]
```
  * There is still mixed use of `hex` and `'0x%x'` to format ids. A single one should be used consistently.

Please commit after fixing the above. 

  * Since `twisted.test.connectionmixins.TCPClientTestsMixin.test_addresses` is being changed, it would be nice to make a single assert with all the relevant data.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

> The tests for the deprecated functions should suppress the warnings from those functions, using something like

Yes, except actually use `twisted.trial.util.suppress` instead.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_15"></a>|@tomprince commented|
|-|-|

[buildbot run](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-unsignedid-5544&num_builds=10)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)|@tomprince removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @tomprince<br>@exarkun set status to `new`|
|-|-|

  1. I don't like the change from `import a, b` to `import a\nimport b\n`.  This is something PEP8 talks about, I guess, but it's a point where Twisted's style diverges.  I don't see any readability reasons to limit an import statement to a single name (we certainly don't come anywhere close to adhering to that for all of our `twisted` imports).  I don't really care if this branch changes a few imports from one style to the other, but I hope you (alas, I have no idea who "you" is, as this branch has been touched by so many) don't go out of your way to make this change in the future.  Change it back if you like, otherwise just keep it in mind for the future.
  1. `test_defer` module docstring is missing an article - "the" would fit nicely before the module name.  Alternatively, drop the trailing _module_.
  1. The change to the test in `twisted/internet/test/test_base.py` makes the assertion worse.  Previously the test was very clear about what string it expected.  Now it relies on "%x" being the correct way to format an integer in hex to construct the string it expects.  String literals are better than constructed strings in this context.  They leave less room for mistakes and misunderstandings.  That said, since there is no `setBuiltinID` and since monkey patching the `id` attribute of the `base` module is pretty gross, maybe this ticket inevitably just makes this test (and others) worse.
  1. in `test_reflectpy3`, `test_classIDStr` seems misnamed to me.  The test exercises `safe_repr`.  I also think the point of the test has nothing to do with class objects, really; it's about the inclusion of the (arbitrary) object id in the error report string.
  1. The actual assertion made by that test makes is pretty weak as well, `assertIn` tells us very little about the string (clearly that's a pre-existing issue, not a problem introduced by this change).
  1. The indentation for the new code in `test_addresses` is all a bit funny.  This is the sort of thing I really wish twistedchecker knew about. :/

Please fix 2, 4, and 6 and file a ticket for 5, then merge.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_19"></a>|@tomprince set status to `closed`|
|-|-|

(In [36856]) Merge deprecate-unsignedid-5544: Deprecate `twisted.python.util.unsignedID` and replace its uses with `id`.

Author: zomux, thijs, tom.prince
Reviewers: tom.prince, exarkun
Fixes: [#5544](https://github.com/twisted/twisted/issues/5544)

twisted.python.util.unsignedID was created to deal with id returning negative
numbers. Python 2.5 and later don't have this problem and id now returns a
positive number always.

Replace all uses of this function in Twisted with id, then deprecate it.
